### DISCUSSION TITLE
C interop today and in the future.

### DISCUSSION THREAD

**Discussion by dabrahams** [hylo-lang/hylo#1122] [Language design]
Unfortunately, calling a C-language function from LLVM is not as simple as creating a trivially-corresponding LLVM declaration for the C function and calling that.  There are [all kinds of irregularities](https://youtu.be/_xAqf-VwaOM?si=RDGE_O_XJx028ZGG&t=1918) in how a given C call should be lowered to LLVM across target architectures and platforms, and those details are currently captured in the Clang codebase.

However, LLVM does make an [informal guarantee of regularity](https://www.youtube.com/watch?v=_xAqf-VwaOM&t=1995s) for calls taking pointer and pointer-sized integer parameters.

Summarizing John McCall from the linked talk,

> When you have an LLVM type that would naturally turn into a bunch of pointers or pointer size integers on standard cpu architectures, the naive lowering… will actually work. So that's… an informal guarantee that you can rely on across platforms 

(John defines **register-filling types** as pointer types and pointer-sized integer types; let's additionally define an **LLVM-trivial** function as any C function whose parameter types are register-filling, and whose return type is register-filling or `void`.)

but he goes on to caution, 

> that guarantee is extremely limited. Don't try to stretch too far… when you're interacting with a[n arbitrary] c declaration  you really need to be using Clang to lower the interaction.

## The future

As the linked talk says, that is exactly what Swift does, and it seems apparent that in order to ingest C headers and be able to use the types and functions declared there efficiently and cleanly from Hylo, we will eventually want to do the same.  That will free us from relying on LLVM's informal guarantees and allow us to call into C from Hylo in the most efficient and straightforward way, without writing any glue code.  It will also give us tools we'll want for building Hylo modules and consuming C modules.  [Swift/C++ interop](https://www.swift.org/documentation/cxx-interop/) should be a big help in this effort.

## Stopgaps for calling C functions

But what do we do in the meantime?  I can see several possible approaches, each building on the previous one:

1. Document LLVM's informal rules, using something like our existing `@ffi` declarations to make LLVM-trivial C functions visible to Hylo.  Since few existing C functions are LLVM-trivial, we would often manually write LLVM-trivial trampoline functions in C and link them into a hylo program.
2. If we  sacrifice the ability to make trampoline-free calls to the few existing C functions that happen to be already LLVM-trivial, we can dispense with `ForeignConvertible`.  Instead, our `@ffi` declarations can simply declare the Hylo signature of a trampoline that uses Hylo's calling convention (which creates only LLVM-trivial functions, because everything is passed and returned by pointer), with the C code of a trampoline dereferencing each pointer argument as necessary to extract C argument values and store the C return type.
2.  Create C++ templates that can easily generate an appropriate trampoline for any arbitrary C function, making trampolines far more straightforward to write.
3. Have the hylo compiler:
    1. generate trampolines in C or C++ source code from `@ffi` declarations
    2. invoke the appropriate compiler to create an object file
    3. automatically link the result into the executable.

How many of these steps to actually pursue before we integrate Clang's source into Hylo is an open question.
 
## Calling Hylo functions from C

In general, function pointers handled by C code are not expected to use Hylo calling conventions, so a C function pointer that calls Hylo must refer to a **callback wrapper**, written in C (or generated using Clang), that marshals argument and return values into/out of a call to Hylo.  In the most general case, the callback wrapper must be written with explicit knowledge of the Hylo function, which must have a mangled name that can be referred to from a C declaration.  In many cases, though, the C signature includes a `void*` referring to “captures” of a notional closure of which the function pointer is the other part. In these cases it may be possible to store the address of the Hylo function to be called inside the captures, and a single callback wrapper for a given C signature can be used with multiple Hylo functions.

I expect many of the details of how this should work to be revealed by the process of manually creating callback wrappers.

**Comment by DibyaXPP** [comment]
Will their be Modern C++ Transpilling? Like I can write Hylo code with VSCODE, Transpile to C++ CPP File & Modules then compile with my old code in Visual Studio with MSVC Compiler. 
I think destructive move is biggest concern for Hylo but i guess their proposal for that in C++ https://github.com/tringi/papers/blob/main/cxx-destructive-move.md

**Reply by dabrahams** [reply]
It's not on the roadmap, but C++ interop is (eventually) very interesting to us and we're not opposed to having transpilation as part of that picture.  At this point we don't have the manpower to work on the design and implementation, but if you'd like to contribute, we could give your designs and PRs some attention.

**Reply by DibyaXPP** [reply]
I will love to help, but i need to understand how the language works first including mutable value semantics. Some more documentation will be really nice but obviously documentation is secondary, first language needs to evolve. 

I love what u gals & guys doing.

**Comment by tothambrus11** [comment]
I wanted to summarize the possible approaches to C interoperability I discovered so far.

## Definitions
Throughout this post, I'll refer to Hylo/Swift/Rust as *native* calling convention, and to C as *foreign* calling convention.
I'll use the terms trivially copyable/movable/relocatable types to mean that these operations of the given type don't involve anything more than a `memcpy`.

## Introduction

Calling arbitrary C functions requires knowledge of the C ABI, which Clang and other LLVM-backed compilers' frontends reimplement. Specifically, the call is influenced by the exact data type and its internal representation. E.g. a [`char` can be passed differently based on whether it's part of a struct](https://godbolt.org/z/Yvsaqzsxh):

```c
struct S { char k; };

char c(struct S fe) {
    return fe.k;
}

char hylo(char c) {
    return c;
}
```
```llvm
define dso_local zeroext i8 @c([1 x i32] %fe.coerce) local_unnamed_addr {
entry:
  %fe.coerce.fca.0.extract = extractvalue [1 x i32] %fe.coerce, 0
  %tmp.coerce.sroa.0.0.extract.trunc = trunc i32 %fe.coerce.fca.0.extract to i8
  ret i8 %tmp.coerce.sroa.0.0.extract.trunc
}

define dso_local noundef zeroext i8 @hylo(i8 noundef returned zeroext %c) local_unnamed_addr {
entry:
  ret i8 %c
}
```

## 1. The Rust approach
Therefore, to call C functions via Clang, we not only would need to embed Clang and use it to generate the calls, we would also need our type system to accurately describe the C type system.

Rust attempted this approach (though without Clang, doing everything on their own), and they have various unsolvable challenges because some C structs cannot be represented accurately in a `#[repr(C)]` Rust struct. A notable case is bit-fields. The intrusive nature of this architecture is demonstrated by the fact that if we wanted to add C interop to our language this way, we would be required to add bitfields, untagged unions, arbitrarily sized integers and who knows what else to get full coverage.

## 2. The Swift approach

Another approach is to ask users to import C types directly from C header files rather than trying to declare them in our language if they want the widest feature coverage. This allows Swift to maintain the Clang AST nodes of C types throughout the AST and type system, so that when emitting calls, they don't need to read the type information from Swift types but directly from the Clang AST nodes.

Including pointers in the AST can be troublesome for serialization, so Swift constructs a stable lookup path that uniquely identifies the C/C++ entitiy (e.g. `MyNameSpace -> MyClass -> MyMethod`), then finds the node again in Clang AST upon deserialization based on this.

### Swift's ABI-adopting thunks

Swift models the ABI of imported functions in their types (`@convention(C)`). Native thunks are only generated on demand, when we want to pass a `@convention(C)` function as a closure to some function requiring a Swift calling conversion closure. C function pointers are exposed as `@convention(C)` empty-environment (thin) closure types, and we can trivially pass an imported C function to another C function that takes a function pointer. Here are the variations between native and foreign callees and arguments:

1. `c_sort(&xs, c_compare(_:_:))`
  - No thunks are used. 
  - Function application is performed according to the C ABI with the help of Clang.

2. `swift_sort(&xs, c_compare(_:_:))`
  - Foreign-to-native thunk of `c_compare` is used.
  - Function application is performed according to the Swift calling convention.

3. `c_sort(&xs, swift_compare(_:_:))`
  - Native-to-foreign thunk of `swift_compare` is used.
  - Function application is performed according to the C ABI with the help of Clang.
 
4. `swift_sort(&xs, swift_compare(_:_:))`
  - No thunks are used.
  - Function application is performed according to the Swift calling convention.

In general, **native-to-foreign thunk generation requires the calle to be statically known**. Without a passed-in environment, a signature-specific thunk cannot know what callee to forward to. Workarounds exist, but are either unsafe & non-portable (JIT compiling new thunks on demand) or not arbitrarily scalable at runtime (temporary allocating an instance from pre-generated pool of universal thunks, reading environment from a corresponding global variable).

Foreign-to-native conversion is always possible because the resulting Swift closure can hold the dynamic c function pointer in the closure's environment.

<img width="1318" height="750" alt="image" src="https://github.com/user-attachments/assets/8e9a2226-de31-4978-ae33-c2aa39e3a2b1" />


Swift also uses Clang's CodeGen to emit inline C functions to the Swift module's LLVM module in which they are used. Overall, the Swift model results in a binary with as low overhead as it gets. However, this comes at the cost of deep coupling with Clang's C++ API throughout the compiler code base.

## 3. The LLVM-Trivial Indirect intermediary approach
What if we added the constraint that Hylo code can only call functions with Hylo calling convention? With this constraint, we can effectively decouple the Hylo compiler from Clang codegen, and use any C compiler to compile the C-specific parts.

Here are the variations for different callees and arguments when using this approach:

1. `c_sort(&xs, c_compare(_:_:))`
  - Pass the c function pointer of `c_compare` to to `c_sort`'s foreign-to-native thunk.
  - `c_sort` is called via Hylo calling convention.
2. `hylo_sort(&xs, c_compare(_:_:))`
  - Pass the foreign-to-native thunk of `c_compare` to a Hylo function.
  - `hylo_sort` is called via Hylo calling convention.
3. `c_sort(&xs, hylo_compare(_:_:))`
  - Pass the native-to-foreign thunk of `hylo_compare` to `c_sort`'s foreign-to-native thunk.
  - `c_sort` is called via Hylo calling convention.
4. `hylo_sort(&xs, hylo_compare(_:_:))`
  - Pass `hylo_compare` without thunks.
  - `hylo_sort` is called via Hylo calling convention.

What this requires:
- Conversion from imported C function to C function pointer requires aquiring the function pointer of the original C function.
- A foreign-to-native thunk is needed when an imported C function is called or conversion to a thin Hylo closure.
- Conversion from Hylo function to C function pointer requires native-to-foreign thunk.

### Indirect calling convention
Without Clang, the Hylo compiler cannot reason about arbitrary calling conventions on its own, except when all parameters are pointers or pointer-sized integers. Because of this, the foreign-to-native and native-to-foreign thunks must be generated with an additional layer of indirection, through an LLVM-trivial thunk.


#### The foreign-to-native conversion happens as follows:
- A C function we want to import is declared and compiled via a C compiler.
- A foreign-to-indirect thunk (calling our C function) is compiled by a C compiler.
- An indirect-to-native thunk (calling the foreign-to-indirect thunk) is compiled by the Hylo compiler.

<img width="2458" height="1104" alt="image" src="https://github.com/user-attachments/assets/85c3d505-68bb-4073-8935-0266b9561daf" />

#### The pointer-collapsing behavior of foreign-to-native conversion
By default, all C pointer parameters are exposed as `UnsafePointer<T>` or `UnsafeMutablePointer<T>`. However, we can recognize that most C functions' pointer parameters correspond to Hylo's `inout`/`sink`/`set`/`set` semantics. To improve the user experience on the Hylo side, the mapping can be programmatically customized.

The following is an example strategy that covers all pointer parameters in Raylib (except ones that are array-like):

```swift
struct RaylibStrategy: BindingStrategy {

  func convertPointerParameter(f: FunctionDecl, p: ParmDecl) -> AccessEffect? {
    let type = p.type!.prettyPrinted(using: f.printingPolicy)

    if f.spelling.hasPrefix("UpdateCamera") && p.spelling == "camera"
      || f.spelling.hasPrefix("Image") && type == "Image *"
      || f.spelling == "GenTextureMipmaps" && type == "Texture2D *"
      || (f.spelling == "UploadMesh" || f.spelling == "GenMeshTangents") && type == "Mesh *"
      || f.spelling == "SetMaterialTexture" && type == "Material *"
      || f.spelling == "SetModelMeshMaterial" && type == "Model *"
      || f.spelling.hasPrefix("Wave") && type == "Wave *"
      || f.spelling == "TextAppend" && type == "int *"
    {
      return .inout
    }

    let setParameterNames = [
      "dataSize", "compDataSize", "outputSize", "frames", "fileSize", "colorCount", "glyphCount",
      "count", "codepointSize", "utf8Size", "materialCount", "animCount",
    ]
    if type == "int *" && setParameterNames.contains(p.spelling) {
      return .set
    }

    return nil
  }
  
}
```


#### The native-to-foreign conversion happens as follows:
- A native Hylo function is declared and defined (with the `@expose` annotation)
- A native-to-indirect thunk (calling the native Hylo function) is compiled by the Hylo compiler.
- An indirect-to-foreign thunk (calling the native-to-indirect thunk) is compiled by a C compiler.

<img width="2458" height="1302" alt="image" src="https://github.com/user-attachments/assets/e8f4a452-ac70-4a9b-9c07-00a680ae25df" />


#### The pointer collapsing behavior of native-to-foreign conversion
Although the native-to-indirect thunk is already a valid C declaration, the indirect-to-foreign thunk can simplify/collapse the API by accepting trivially relocatable parameters and return values by copy. `inout` and `set` parameters automatically opt out of collapsing and get passed by mutable pointer. In addition, we can annotate `let` and `sink` parameters and the return type with `@indirect` to have them exposed as a pointer parameter in the C API. Non-trivially relocatable types cannot be passed by copy in C and must always use the `@indirect` convention.

Here are some examples:
<img width="1620" height="1462" alt="image" src="https://github.com/user-attachments/assets/434040c3-a421-427d-91b8-5e0288cdab2d" />

While the entities in this approach are quite different (`CFunctionPointer<...>` instead of `@convention(C) [](...) -> T { ... }`), the general expressivity of the first approach is preserved.

<img width="992" height="636" alt="image" src="https://github.com/user-attachments/assets/ced0bcdc-a532-45da-b211-4c9f97d8c1e6" />


### Type Layout and Data Member Interoperability (C -> Hylo)
While the Hylo type system doesn't have perfect knowledge about the internal layout of the struct, we can declare an opaque struct with the correct size and alignment, so that we can correctly allocate and pass around the value by copy.

<img width="1380" height="334" alt="image" src="https://github.com/user-attachments/assets/7671d235-9392-4491-b9a4-badc03d71750" />

Furthermore, we can declare accessors in Hylo (computed properties) that can be implemented in C or Hylo. Here are some alternatives that we can work with:

- **Trivially Copyable, Movable member:**
  ```c
  void get_Point_x(Point const* self, float* result);
  void set_Point_x(Point const* self, float const* x, float* result);
  ```

- **Non-trivially Copyable/Movable member:**
  Pass the required witnesses.
  ```c
  void get_Point_x(Copyable<float>* w, Point const* self, float* result);
  void set_Point_x(Movable<float>* w, Point const* self, float const* x, float* result);
  ```

- **Regular Subscripts as accessors**
  We can avoid passing the witnesses, by inversion of control.
  ```
  void Point_x_ramp(Point const* self, FUN_PTR<(float* x, ...) → void>, ...);
  void Point_x_slide(Point const* self, ProjectionFrame*);
  ```

- **Addressors**
  A specialized calling convention of subscript can be used when we know a value is backed by a stored property at a correctly aligned place. In this case, we can have a function accessor returning a pointer to it:
  ```c
  float* Point_x(Point const* self);
  ```
  Almost everything can be an addressor except packed struct members and bitfields.

- **Addressors - Inlined In Hylo**
  Address computation can be also inlined in a Hylo shim, capturing the C struct’s layout. 
  ```swift
  struct Point {
    ...
    subscript y() → Float32 {
      yield self.raw_address()
        .offset_by(4)
        .unsafe_as<Float32>[]
    }
  }
  ```

C unions and flexible array members can similarly be exposed with special accessors. Flexible array members by default are not `Movable`/`Copyable`. They get a special static initializer subscript that can allocate an instance with given FAM element count.

The question of when to use which kind of accessor strategy is up to discussion. I see 2 main approaches:
- always use the simplest and most efficient ABI (inlined addressors when possible, getters/setters for trivially relocatable bitfields/packed members, and subscripts for non-trivially relocatable members.
- always use a consistent, general ABI for simplicity (regular subscripts).

### Type Layout and Data Member Interoperability (Hylo -> C)
The backwards mapping can be more or less the same as C->Hylo mapping, except that the accessors are typically implemented in Hylo. For C code (and other languages) it is generally less convenient to consume Hylo subscripts, so addressors and getters/setters would be preferred when possible.

## Discussion
I think we can safely rule out the Rust approach to interoperability. Having to form the type system of the new language such that every C language construct can be mapped is too intrusive of a design, and leads to incompleteness. 

Historically, C library developers were discouraged from using specific language features in their public API so that it can be easier to consume by FFI tools. While introducing higher-fidelity interop, like Swift, we unlock some capabilities and expressive power in the C language, and generally create a better cross-language experience. However, for the time being, library developers will keep being discouraged from using bitfields and packed structs in ABIs, because it's not just Hylo and Swift that they need to support.

Now we are left with the choice between Swift's approach (integrating Clang for codegen), and the indirect thunk based approach. The former is likely more intrusive to integrate, and introduces an additional dependency. However, it is the most efficient solution, as it can avoid function call overheads as much as possible. The latter approach only requires an arbitrary C compiler and an external build tool (`hc-bindgen`) that can generate the necessary C source code thunks. 

The Clang-based approach can also naturally lead to optimal code size, as thunks are emitted lazily. In the indirect approach, extracting the information for `hc-bindgen` about which thunks to actually generate is a bit more complicated from a build system standpoint. We would need to analyze the Hylo IR after dead code elimination and see which thunks are ever referenced. But for now, it is also acceptable to eagerly generate all thunks, which is quite simple.

**Reply by dabrahams** [reply]
- Nit: I don't understand what advantage the word “foreign” has over “C,” and it seems like it could be confusing as soon as we start to discuss interop with other languages than C, where the same issues/reasoning may not apply
- The introduction of the Rust approach is confusing to the point where I don't know what you mean by the phrase.  If it's not exactly “embedding Clang,” then what is it, exactly?  The rest is based on this poorly-defined phrase.  My best guess is it means, “ask users to redeclare C types in the native language in a way compatible with their C declarations.”
- You seem to take as a premise that “full coverage” is a requirement.  I'm not sure that's true, especially not immediately.
- The Swift approach is also rather unclear.  Except for the first sentence, everything in the first two paragraphs seems like implementation detail that doesn't belong.
   - You use the term “native thunk” without defining it.  I can sort of guess, but it's not perfectly clear.
   - “calling conversion” should be “calling convention”
   - In  the `c_sort` examples I don't see how clang is involved in function application.  The code for `c_sort` was presumably already compiled somewhere, possibly by a Rust compiler or GCC.
   - In fact the whole list looks redundant with the other information you've given.  Isn't it sufficient to replace it, and the first paragraph of the thunk section with, “Swift implicitly generates ‘thunk’ function wrappers when C function pointers are passed as native closures (or vice-versa)?”

I'm starting to get the feeling that I'm being asked to read past massive amounts of needless verbiage.  Did you use AI to generate (drafts of) this text?

**Reply by dabrahams** [reply]
In this section

> - A C function we want to import is declared and compiled via a C compiler.
> - A foreign-to-indirect thunk (calling our C function) is compiled by a C compiler.
> - An indirect-to-native thunk (calling the foreign-to-indirect thunk) is compiled by the Hylo compiler.

The first bullet seems needless.

It only makes sense to me if I take a counterintuitive view of the word “to,” such that a foreign-to-indirect thunk creates an indirect function that makes a call **to** a foreign function, and interpret the word “indirect” as “LLVM-trivial”.  Sure, you are turning a foreign API **into** an LLVM-trivial API, but it should just be 

> - A C thunk with an LLVM-trivial API calling the C function is compiled by a C compiler.
> - A thunk that calls the LLVM-trivial thunk is compiled by the Hylo compiler.

**Reply by tothambrus11** [reply]
Thanks for the feedback, very good observations. 

> Nit: I don't understand what advantage the word “foreign” has over “C,” and it seems like it could be confusing as soon as we start to discuss interop with other languages than C, where the same issues/reasoning may not apply

Swift used this terminology in its compiler code base, and I wanted to be consistent with it. But indeed, it's not helping here, so I can replace with the clearer C-to-Hylo style terminology, and explain each of these.

> In fact the whole list looks redundant with the other information you've given. Isn't it sufficient to replace it, and the first paragraph of the thunk section with, “Swift implicitly generates ‘thunk’ function wrappers when C function pointers are passed as native closures (or vice-versa)?”

Yes, thank you, that's exactly the case. I can collapse the examples and use your phrasing.

> You seem to take as a premise that “full coverage” is a requirement. I'm not sure that's true, especially not immediately.

Yes, my goal is not full coverage (especially not immediately). Instead, my goal is that aiming for more coverage should scale without affecting Hylo's type system. E.g. we shouldn't need to add language features like untagged unions just to cover interoperability. To my knowledge, almost all of C's features can be emulated through a struct with opaque storage with accessors on it.

> In the c_sort examples I don't see how clang is involved in function application. The code for c_sort was presumably already compiled somewhere, possibly by a Rust compiler or GCC.

When calling a C function, Swift calls it using the C calling convention. To set up the parameters and stack correctly, Swift uses Clang to perform this according to the target ABI. This is the bit that LLVM is not prepared to do on its own and where frontends must still maintain target-specific knowledge.

> I'm starting to get the feeling that I'm being asked to read past massive amounts of needless verbiage. Did you use AI to generate (drafts of) this text?

Not at all, I've been just sitting over this problem for too much time, and the text grew long when I wrote it down. I will do some passes of review and improve the wording. I believe I have a solid understanding of the topic, though I can't claim to not have any gaps, so I appreciate any suggestions.


Thanks also for the rest of wording suggestions, I'll apply them asap.

**Reply by dabrahams** [reply]
OK, but the function application I thought of was the one where `c_sort` calls its argument.  Why not simply say “to call `c_sort`?”
